ts: introduce a new 1 minute poller to write into tsdb

jasonlmfong opened the pull request:
This change adds a new 1 minute resolution and the machinery around rollup and storage.
Also adds a new poller at server start-up time, this poller currently performs a no-op every 1 minute.

Epic: CRDB-55079
Release: None

dhartunian reviewed (changes requested):
> @dhartunian reviewed 6 of 11 files at r1, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/157226)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained

___
*[`-- commits` line 5 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/157226#-OeHhTXA5Yv5W7suFTD_:-OeHhTXA5Yv5W7suFTDa:bvfxo41):*
it's unclear here what the new poller is doing in this PR. is it a no-op for now? is it recording something?
___
*[`pkg/ts/db.go` line 175 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/157226#-OeI8u1vEwv9eYKtmZGB:-OeI8u1vEwv9eYKtmZGC:b-lixqg6) ([raw file](https://github.com/cockroachdb/cockroach/blob/b069ba1d823b844c59723c8c6a7e730abc5f62ef/pkg/ts/db.go#L175)):*
> ```Go
> func (db *DB) PollChildMetricsSource(
> 	ambient log.AmbientContext,
> 	source ChildMetricsDataSource,
> ```

What do we gain by creating a custom poller just for child metrics? why can't we just call `PollSource` twice with different args? It would be nice to share the poller code.

Could you, for instance, add a boolean arg to `GetTimeSeriesData(childMetrics)` which asks for child metrics?
Or maybe we could add a method to MetricsRecorder which constructs a metric source that's just child metrics and use that.

I know this implementation keeps the existing code path untouched, which is kinda nice. But I think adding a single boolean arg is a reasonable risk to take in this case, and would be straightforward to backport.
___
*[`pkg/server/status/recorder.go` line 113 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/157226#-OeI8C3q0UIYl9O035qW:-OeI8C3r1beMeOD2p0ZM:b-a1o6cl) ([raw file](https://github.com/cockroachdb/cockroach/blob/b069ba1d823b844c59723c8c6a7e730abc5f62ef/pkg/server/status/recorder.go#L113)):*
> ```Go
> // ChildMetricsTimeSeriesEnabled controls whether to collect high-cardinality child metrics
> // into the time series database. This is separate from ChildMetricsEnabled which controls
> // Prometheus exports, allowing independent control of child metrics collection vs export.
> ```

Add an extra note here that this setting is meant to enable debugging of changefeeds and should not be considered functionality to expand support for.
___
*[`pkg/server/status/recorder.go` line 118 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/157226#-OeHhsbbE1w4mYAGvMZS:-OeHhsbbE1w4mYAGvMZT:b8c8uik) ([raw file](https://github.com/cockroachdb/cockroach/blob/b069ba1d823b844c59723c8c6a7e730abc5f62ef/pkg/server/status/recorder.go#L118)):*
> ```Go
> 	"enables the collection of high-cardinality child metrics into the time series database",
> 	false,
> 	settings.WithPublic)
> ```

We should discuss whether this metric is "public" and documented or not. I think it shouldn't be.
___
*[`pkg/server/status/recorder.go` line 519 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/157226#-OeI80gL7US3ic8R9wAc:-OeI80gL7US3ic8R9wAd:b-eemw9j) ([raw file](https://github.com/cockroachdb/cockroach/blob/b069ba1d823b844c59723c8c6a7e730abc5f62ef/pkg/server/status/recorder.go#L519)):*
> ```Go
> // interface for high-cardinality child metrics collection.
> func (mr *MetricsRecorder) GetChildMetricsData() []tspb.TimeSeriesData {
> 	// Check if child metrics time series collection is enabled via cluster setting
> ```

nit: I'd remove comments like this. It's obvious what the code is doing by reading at it.
___
*[`pkg/server/status/recorder.go` line 524 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/157226#-OeI7uNi0TdBq3MoKPIf:-OeI7uNi0TdBq3MoKPIg:b-m5liih) ([raw file](https://github.com/cockroachdb/cockroach/blob/b069ba1d823b844c59723c8c6a7e730abc5f62ef/pkg/server/status/recorder.go#L524)):*
> ```Go
> 	}
> 
> 	return nil // TODO (jasonlmfong): to be implemented
> ```

nit: TODO format does not have a space between `TODO` and `(`.
___
*[`pkg/ts/pruning_test.go` line 486 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/157226#-OeICfn07_zyHfSPVFwo:-OeICfn07_zyHfSPVFwp:byga7ci) ([raw file](https://github.com/cockroachdb/cockroach/blob/b069ba1d823b844c59723c8c6a7e730abc5f62ef/pkg/ts/pruning_test.go#L486)):*
> ```Go
> 	// 2 metrics * 2 sources * 3 resolutions * 1 key = 12
> 	// 2 metrics * 2 sources * 2 resolutions * 1 key = 8
> 	// note that the 10s data and 1m data rolls into the same 30m rollup resolution.
> ```

Hm this makes me wonder...if you have the same data in 10s and 1m and you do the rollup, does one overwrite the other in the 30m resolution? Below, it seems that after pruning we still have 3 resolutions. Is that expected?


<!-- Sent from Reviewable.io -->

jasonlmfong reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/157226)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @dhartunian)

___
*[`pkg/server/status/recorder.go` line 113 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/157226#-OeI8C3q0UIYl9O035qW:-OeIOMmwE4W7h74sXCS2:bb7oryy) ([raw file](https://github.com/cockroachdb/cockroach/blob/b069ba1d823b844c59723c8c6a7e730abc5f62ef/pkg/server/status/recorder.go#L113)):*
<details><summary><i>Previously, dhartunian (David Hartunian) wrote…</i></summary><blockquote>

Add an extra note here that this setting is meant to enable debugging of changefeeds and should not be considered functionality to expand support for.
</blockquote></details>

sure
___
*[`pkg/ts/db.go` line 175 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/157226#-OeI8u1vEwv9eYKtmZGB:-OeIOcb-5z4SSplJLEh1:bly3cvs) ([raw file](https://github.com/cockroachdb/cockroach/blob/b069ba1d823b844c59723c8c6a7e730abc5f62ef/pkg/ts/db.go#L175)):*
<details><summary><i>Previously, dhartunian (David Hartunian) wrote…</i></summary><blockquote>

What do we gain by creating a custom poller just for child metrics? why can't we just call `PollSource` twice with different args? It would be nice to share the poller code.

Could you, for instance, add a boolean arg to `GetTimeSeriesData(childMetrics)` which asks for child metrics?
Or maybe we could add a method to MetricsRecorder which constructs a metric source that's just child metrics and use that.

I know this implementation keeps the existing code path untouched, which is kinda nice. But I think adding a single boolean arg is a reasonable risk to take in this case, and would be straightforward to backport.
</blockquote></details>

I can try that
___
*[`pkg/ts/pruning_test.go` line 486 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/157226#-OeICfn07_zyHfSPVFwo:-OeIP-hR7qGa3ZPmIH_M:b8ak420) ([raw file](https://github.com/cockroachdb/cockroach/blob/b069ba1d823b844c59723c8c6a7e730abc5f62ef/pkg/ts/pruning_test.go#L486)):*
> does one overwrite the other in the 30m resolution

This I was assuming that they are "rolled together" but you might be right in that it is overwriting

> Below, it seems that after pruning we still have 3 resolutions. Is that expected?

yes that is expected, it's the 3 resolutions 1ns, 10s, 1m that stay
the 2 that were pruned are 30m and 50ns (roll ups that are too old)


<!-- Sent from Reviewable.io -->

jasonlmfong reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/157226)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @dhartunian)

___
*[`pkg/ts/pruning_test.go` line 486 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/157226#-OeICfn07_zyHfSPVFwo:-OeIRvntBLOS5w963Xb7:bc7yeo8) ([raw file](https://github.com/cockroachdb/cockroach/blob/b069ba1d823b844c59723c8c6a7e730abc5f62ef/pkg/ts/pruning_test.go#L486)):*
> does one overwrite the other in the 30m resolution

I believe it is aggregated together in this test
in any case, I don't think we will see this happen in the wild as long as the new 1m resolution is only used for metrics with extra labels as names


<!-- Sent from Reviewable.io -->

dhartunian reviewed (approved):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/> just some nits. feel free to merge once those are fixed.

> @dhartunian reviewed 2 of 2 files at r3.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/157226#-:-OeMZozc27OQrbNrqeqk:b-4e00i0)* status: <span>:shipit:</span> complete! 1 of 0 LGTMs obtained

___
*[`-- commits` line 5 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/157226#-OeMZ0dJ9WSpVjmgNIq0:-OeMZ0dJ9WSpVjmgNIq1:b-l5e54e):*
nit: can you check the wrapping on your commit message.
___
*[`pkg/ts/db.go` line 129 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/157226#-OeMZf5P8c5vu-Xh9nBK:-OeMZf5P8c5vu-Xh9nBL:b-1urc1k) ([raw file](https://github.com/cockroachdb/cockroach/blob/ad3ece095266755126d2c457b0d4bd00d7463b98/pkg/ts/db.go#L129)):*
> ```Go
> 	r           Resolution
> 	stopper     *stop.Stopper
> 	childMetrics bool
> ```

nit: can you add a docstring to this arg since the behavior it triggers will be a bit "special"


<!-- Sent from Reviewable.io -->

angles-n-daemons reviewed (approved):
nice, added a non-blocking suggestion
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/157226#-:-OeOBD2O6QBXk4i2jlqG:b-h3rvl2)* status: <span>:shipit:</span> complete! 1 of 0 LGTMs obtained (and 1 stale) (waiting on @dhartunian and @jasonlmfong)

___
*[`pkg/server/status/recorder.go` line 116 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/157226#-OeO7WKfAub6EovBtQIA:-OeO7WKfAub6EovBtQIB:bfryh1p) ([raw file](https://github.com/cockroachdb/cockroach/blob/94bf51623529564ad3e96ca6728ab9b0127c79c5/pkg/server/status/recorder.go#L116)):*
> ```Go
> // This setting enables debugging of changefeeds and should not be considered functionality
> // to expand support for.
> var ChildMetricsTimeSeriesEnabled = settings.RegisterBoolSetting(
> ```

suggestion: this might read a little better as `ChildMetricsStorageEnabled`, the phrase timeseries feels ambiguous despite referring to tsdb.


<!-- Sent from Reviewable.io -->

jasonlmfong committed:
ts: introduce a new 1 minute poller to write into tsdb

This change adds a new 1 minute resolution and the machinery around rollup and storage. This change also adds a new poller at server start-up time, this poller currently performs a no-op every 1 minute.

Epic: CRDB-55079
Release: None
